### PR TITLE
Feature/two phase

### PR DESCRIPTION
**Pull request by biniyam-mt**
Implemented delayed lift for function bodies until all declarations are lifted.

### CODE REVIEW

File: lib/AST.cpp [hunk 1]
```diff
@@ -25,6 +27,7 @@ namespace pillar
 {
   namespace ast
   {
+    std::vector<std::function<void(void)>> *AST::lift_queue = new std::vector<std::function<void(void)>>();
```
Comment (pgoodman) [inline lib/AST.cpp:30]: make this a member/field of the `AST` class. You don't need to make it heap-allocated either, just `std::vector<std::function<void(void)>> lift_queue;` will do.

File: lib/AST.h [hunk 2]
```diff
@@ -61,6 +61,7 @@ namespace pillar
       const std::shared_ptr<mlir::Operation> module;
       const mlir::DataLayout dl;
       const NameProvider np;
+      static std::vector<std::function<void(void)>> *lift_queue;
```
Comment (pgoodman) [inline lib/AST.h:64]: You don't need to make it `static` or heap-allocated either, just `std::vector<std::function<void(void)>> lift_queue;` will do.

File: lib/AST.cpp [hunk 3]
```diff
@@ -62,8 +65,45 @@ namespace pillar
...
-                                                                             { (void)ast->LiftFuncOp(tu, tu, std::move(op)); });
+
+      ///////
+      for (mlir::Operation &op : moduleOp.getBody()->getOperations())
+      {
+        // Print the operation name.
+        llvm::TypeSwitch<mlir::Operation *>(&op)
+            .Case([&](vast::hl::FuncOp)
+                  {
+                                              // std::cout << op.getName().getStringRef().str() << "\n";
```
Comment (pgoodman) [inline lib/AST.cpp:76]: Remove commented out logging.

File: lib/ASTDecl.cpp [hunk 4]
```diff
@@ -16,9 +16,13 @@ namespace pillar
                                          clang::DeclContext *ldc,
                                          vast::hl::FuncOp func)
     {
+
+      std::string functionName = np.FunctionName(func);
```
Comment (pgoodman) [inline lib/ASTDecl.cpp:20]: code style: variables could be `snake_case`.

File: lib/ASTDecl.cpp [hunk 5]
```diff
@@ -60,27 +63,43 @@ namespace pillar
         assert(body.getBlocks().empty());
         return func_decl;
       }
+      AST::lift_queue->emplace_back([&body, func_decl, functionName, this](void)
```
Comment (pgoodman) [inline lib/ASTDecl.cpp:66]: Remove `functionName`, not needed in here. You can shorten this as: `[&body, =, this]`

Comment (pgoodman) [reply lib/ASTDecl.cpp:66]: Extract this lambda into a local var; it'll keep the indentation down. E.g.

```c++
auto lift_body = [&body, =, this] (void) {

};
lift_queue.emplace_back(std::move(lift_body));
```

File: lib/ASTDecl.cpp [hunk 6]
```diff
@@ -60,27 +63,43 @@ namespace pillar
...
+                                          if (!ElideFromCompoundStmt(op, stmt))
+                                          {
+                                            body_stmts.emplace_back(stmt);
+                                          }
+                                        }
+                                      }
+
+                                      clang::FPOptionsOverride fpo;
+                                      clang::CompoundStmt *body_stmt = clang::CompoundStmt::Create(
+                                          ctx, body_stmts, fpo, clang::SourceLocation(), clang::SourceLocation());
```
Comment (pgoodman) [inline lib/ASTDecl.cpp:85]: I think there's a `kEmptyLoc` that you can use in place of `clang::SourceLocation()`. You should see it used elsewhere.

File: lib/NameProvider.h [hunk 7]
```diff
@@ -9,8 +9,11 @@ class NameProvider
 public:
     virtual ~NameProvider(void) = default;
```
Comment (pgoodman) [inline lib/NameProvider.h:10]: Move these includes to the default .cpp file, and instead forward declare all the various ops inside of the `vast::hl` namespace in this header.

**Review by pgoodman — CHANGES_REQUESTED** [review]

File: lib/AST.h (hunk 2 above)
Comment (biniyam-mt) [reply lib/AST.h:64]: we use lift_queue in a static member function CreateFromModule, this was the only way I was able to use it.

**Pull request merged**